torch/csrc/jit/codegen/cuda/lower_alias_memory.cpp

```diff
@@ -0,0 +1,232 @@
+    for (auto expr : exprs) {
+      handle(expr);
+    }
+
+    // Iterate over candidates to find match
+    for (auto tv : candidate_alias_tv_) {
+      TORCH_INTERNAL_ASSERT(
+          map_tv_to_origin_expr_.find(tv) != map_tv_to_origin_expr_.end());
+
+      const auto& expr = map_tv_to_origin_expr_[tv];
+      const auto output = expr->output(0)->as<TensorView>();
+
+      TORCH_INTERNAL_ASSERT(
+          map_tv_to_allocations_.find(output->name()) !=
+          map_tv_to_allocations_.end());
+
+      auto output_alloc = map_tv_to_allocations_[output->name()];
+
+      auto input_alloc =
+          findCompatibleInputAllocate(print_size(output_alloc), expr);
+      if (input_alloc != nullptr) {
+        // std::cout << "Alias Match\t" << output->getMemoryType() << std::endl;
+        output_alloc->setAlias(input_alloc);
+      }
+    }
+  }
+
+ private:
+  // Check if we are a Pointwise TensorView op.
+  bool isPwiseTVOp(const Expr* expr) {
+    // Ignore set operations
+    if (expr->outputs().size() == 1 && ir_utils::isTV(expr->output(0)) &&
+        ((expr->getExprType().value() == ExprType::UnaryOp &&
+          expr->as<UnaryOp>()->getUnaryOpType() != UnaryOpType::Set) ||
+         expr->getExprType().value() == ExprType::BinaryOp ||
+         expr->getExprType().value() == ExprType::TernaryOp))
+      return true;
+    return false;
+  }
+
+  // Get string representation of Allocate size
+  std::string print_size(kir::Allocate* alloc) {
+    printer_.print_inline(alloc->size());
+    std::string size_str = size_ss_.str();
+    size_ss_.str(std::string());
+    return size_str;
+  }
+
+  // Find an Input Allocate that is compatible with the Output Allocate
+  kir::Allocate* findCompatibleInputAllocate(
+      const std::string& output_size_str,
+      Expr* expr) {
+    // Stop searching if current op is not point-wise
+    if (!isPwiseTVOp(expr)) {
+      return nullptr;
+    }
+
+    const auto& expr_inputs_iter =
+        ir_utils::filterByType<TensorView>(expr->inputs());
+
+    std::vector<TensorView*> expr_inputs(
+        expr_inputs_iter.begin(), expr_inputs_iter.end());
+
+    for (const auto input : expr_inputs) {
+      auto input_alloc = map_tv_to_allocations_[input->name()];
+
+      // input_allocation == nullptr implies that input_tv is a fusion input.
+      if (input_alloc != nullptr) {
+        // If input is a candidate TensorView
+        //          input allocation has the same size as output allocation
+        //          is not used after this op:
+        // then alias output Allocate to input Allocate
+        if (candidate_alias_tv_.find(input) != candidate_alias_tv_.end() &&
+            output_size_str == print_size(input_alloc) &&
+            map_tv_to_last_usage_[input] <= map_expr_to_pos_[expr]) {
+          return input_alloc;
+        }
+      }
+    }
+
+    // Assume the first argument contains the primary variable
+    // Follow path along point-wise operations
+    if (!expr_inputs.empty()) {
```

Comment:
Are you attempting to recurse on inputs to find a buffer that you can reuse? Is it necessary at this point? Why only check the first input?

Reply:
Yes. 

Here is an example in Softmax. We can handle the subtraction and exp in the same for loop nest. However, T1 won't be the same size as T0 and T2, so aliasing will fail.
```
// Softmax - y = exp(x - max)
T1 = T0 - max
T2 = exp(T2)
```

As for the first argument, I was focusing on in-place operations. e.g. `exp(x)`, `x += 10`, `clamp(x, -10, 10)`
The first argument is the variable that is being overritten.